src/QA/SoftMocks.php

```diff
@@ -949,16 +949,60 @@ public static function callFunction($namespace, $func, $params)
         return call_user_func_array($func, $params);
     }
 
-    public static function callNew($class, $args)
+    public static function callNew($class, $args, $scope)
     {
         if (isset(self::$new_mocks[$class])) {
             return call_user_func_array(self::$new_mocks[$class], $args);
         }
 
         $Rc = new \ReflectionClass($class);
-        return $Rc->newInstanceArgs($args);
+        $Constructor = $Rc->getConstructor();
+
+        if ($Constructor && self::isConstructorCanBeForced($Constructor, $scope)) {
+            $instance = $Rc->newInstanceWithoutConstructor();
+            $Constructor->setAccessible(true);
+            $Constructor->invokeArgs($instance, $args);
+            $Constructor->setAccessible(false);
+        } else {
+            $instance = $Rc->newInstanceArgs($args);
+        }
+
+        return $instance;
     }
 
+    private static function isConstructorCanBeForced(\ReflectionMethod $constructor, $scope)
+    {
+        if ($constructor->isPublic()) {
+            return false;
+        }
+
+        $class = $constructor->getDeclaringClass();
+        if ($constructor->isPrivate() && $class->getName() === $scope) {
+            return true;
+        }
+
+        if ($constructor->isProtected() && self::isSubclassOf(new \ReflectionClass($scope), $class)) {
```

Comment:
I am not really sure that it is how PHP works :), athough it seems reasonable, of course.

Does the following work?

```
class A { protected function __construct() { } }
class B extends A { public static function createA() { return new A; } }

B::createA();
```

Note that I specified A directly instead of using "parent", for example.
It is very fragile code in a sense that it tries to emulate PHP instead of using it directly :). If behaviour changes in new PHP versions, Soft Mocks will not show it.

I will try to check if the code you provided works properly in all cases that I can find in our badoo code and then probably we could include this functionality, although I would suggest to make it disabled by default.